src/app/services/myinfo/myinfo.service.ts

```diff
@@ -0,0 +1,315 @@
+import {
+  IPersonBasic,
+  IPersonBasicRequest,
+  Mode as MyInfoClientMode,
+  MyInfoGovClient,
+} from '@opengovsg/myinfo-gov-client'
+import Bluebird from 'bluebird'
+import fs from 'fs'
+import { cloneDeep } from 'lodash'
+import mongoose from 'mongoose'
+import { errAsync, ok, okAsync, Result, ResultAsync } from 'neverthrow'
+import CircuitBreaker from 'opossum'
+
+import { IMyInfoConfig } from '../../../config/feature-manager'
+import { createLoggerWithLabel } from '../../../config/logger'
+import {
+  Environment,
+  IFieldSchema,
+  IHashes,
+  IMyInfoHashSchema,
+  MyInfoAttribute,
+} from '../../../types'
+import { DatabaseError } from '../../modules/core/core.errors'
+import { ProcessedFieldResponse } from '../../modules/submission/submission.types'
+
+import getMyInfoHashModel from './myinfo_hash.model'
+import {
+  CircuitBreakerError,
+  FetchMyInfoError,
+  HashDidNotMatchError,
+  HashingError,
+  MissingHashError,
+} from './myinfo.errors'
+import { IPossiblyPrefilledField } from './myinfo.types'
+import {
+  compareHashedValues,
+  getMyInfoValue,
+  hashFieldValues,
+  isFieldReadOnly,
+} from './myinfo.util'
+
+const logger = createLoggerWithLabel(module)
+const MyInfoHash = getMyInfoHashModel(mongoose)
+
+const MYINFO_STG_PREFIX = 'STG2-'
+const MYINFO_PROD_PREFIX = 'PROD2-'
+const MYINFO_DEV_BASE_URL = 'http://localhost:5156/myinfo/v2/'
+
+export class MyInfoService {
+  #myInfoClientBreaker: CircuitBreaker<[IPersonBasicRequest], IPersonBasic>
+  #spCookieMaxAge: number
+
+  /**
+   *
+   * @param myInfoConfig Environment variables including myInfoClientMode and myInfoKeyPath
+   * @param nodeEnv The node environment: development, production or test
+   * @param realm The realm to be passed to MyInfoGovClient
+   * @param singpassEserviceId
+   * @param spCookieMaxAge Validity duration of the SingPass cookie
+   */
+  constructor({
+    myInfoConfig,
+    nodeEnv,
+    realm,
+    singpassEserviceId,
+    spCookieMaxAge,
+  }: {
+    myInfoConfig: IMyInfoConfig
+    nodeEnv: Environment
+    realm: string
+    singpassEserviceId: string
+    spCookieMaxAge: number
+  }) {
+    this.#spCookieMaxAge = spCookieMaxAge
+
+    const { myInfoClientMode, myInfoKeyPath } = myInfoConfig
+    let myInfoGovClient: MyInfoGovClient
+    const myInfoPrefix =
+      myInfoClientMode === MyInfoClientMode.Staging
+        ? MYINFO_STG_PREFIX
+        : MYINFO_PROD_PREFIX
+    if (nodeEnv === Environment.Prod) {
+      myInfoGovClient = new MyInfoGovClient({
+        realm,
+        singpassEserviceId,
+        privateKey: fs.readFileSync(myInfoKeyPath),
+        appId: myInfoPrefix + singpassEserviceId,
+        mode: myInfoClientMode,
```

Comment:
Can simplify this since most of it is the same?

```ts
const myInfoGovClient = new MyInfoGovClient({
        realm,
        singpassEserviceId,
        privateKey: fs.readFileSync(myInfoKeyPath),
        appId: myInfoPrefix + singpassEserviceId,
        mode: nodeEnv === Environment.Prod ? myInfoClientMode : MyInfoClientMode.Dev,
        baseUrl: nodeEnv === Environment.Prod ? undefined : MYINFO_DEV_BASE_URL
})
```

not sure if that makes it clearer or not though haha

Reply:
hm, arguable! personally I think the existing way is cleaner just because of the `baseUrl` line. the code you wrote won't work because `baseUrl` isn't a constructor argument, it's a class member that we're modifying from outside.

Reply:
Got it